### PR TITLE
feat(SCT-1771): Add worker to allocation

### PR DESCRIPTION
**Pull request by mapleleafjack**
This PR needs to be merged AFTER https://github.com/LBHackney-IT/lbh-social-care-frontend/pull/884, as it contains some parts made on that PR. Use this link to compare the changes only this PR introduces: https://github.com/LBHackney-IT/lbh-social-care-frontend/compare/add-team-allocation...add-worker-allocation

**What**  
This PR adds a new "Allocate a worker" form.
The URL to get to the page is `http://dev.hackney.gov.uk:3000/residents/{id}/allocations/{allocationId}/allocateworker?teamId={teamId}&teamAllocationStartDate={allocation_start_date}`

Example working URL in Staging:
`http://dev.hackney.gov.uk:3000/residents/34/allocations/123/allocateworker?teamId=15&teamAllocationStartDate=12-31-2021`

Screenshot 2022-03-21 at 13 10 00

**Why**  
This form is needed to add a worker to an allocation that has a team but not a worker.

**Anything else?**

- Have you added any new third-party libraries? X
- Are any new environment variables or configuration values needed? X
- Anything else in this PR that isn't covered by the what/why? X

### CODE REVIEW

File: components/Allocations/AddWorkerAllocation/AddWorkerAllocation.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  allocationId,
+  teamId,
+  teamAllocationStartDate,
+}: Props): React.ReactElement => {
+  const [postError, setPostError] = useState<boolean | null>();
+  const [postLoading, setPostLoading] = useState<boolean>(true);
+  const [worker, setWorker] = useState<number | null>();
+  const [dateValidation, setDateValidation] = useState<boolean>(false);
+  const [allocationDate, setAllocationDate] = useState<Date>(
+    new Date(Date.now())
```
Comment (acodeninja) [inline components/Allocations/AddWorkerAllocation/AddWorkerAllocation.tsx:33]: This is superfluous.

```javascript
new Date()
// Wed Mar 23 2022 13:10:47 GMT+0000 (Greenwich Mean Time)
new Date(Date.now())
// Wed Mar 23 2022 13:10:54 GMT+0000 (Greenwich Mean Time)
```

File: cypress/integration/team_allocations.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import { AuthRoles } from '../support/commands';
+import { format } from 'date-fns';
+
+describe('Worker / team allocation', () => {
+  describe('As a user in the Admin Dev group', () => {
+    it('Loads correcty the "add an allocation" page', () => {
```
Comment (acodeninja) [inline cypress/integration/team_allocations.ts:6]: Probably not an issue but you might want to `correcty` this.

File: cypress/integration/team_allocations.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        )}/allocations/${Cypress.env(
+          'ALLOCATION_ID'
+        )}/allocateworker?teamId=${Cypress.env(
+          'TEAM_ID'
+        )}&teamAllocationStartDate=12-31-2022`,
+        AuthRoles.AdminDevGroup
+      );
+
+      cy.contains('Allocate a worker').should('exist');
+      cy.contains('Mike Gallagher').should('exist');
```
Comment (acodeninja) [inline cypress/integration/team_allocations.ts:66]: Not an issue, but I'm not sure about using real people's names in tests.

File: lib/allocatedWorkers.spec.ts [hunk 4]
```diff
@@ -205,4 +207,27 @@ describe('allocatedWorkersAPI', () => {
       expect(data).toEqual({ allocations: ['foo'] });
     });
   });
+
+  describe('addWorkerAllocation', () => {
+    it('should work properly', async () => {
```
Comment (acodeninja) [inline lib/allocatedWorkers.spec.ts:212]: Its not a massive issue, but you might want to make the name of this test more meaningful than "it works".

File: pages/api/residents/[id]/allocations/index.ts [hunk 5]
```diff
@@ -49,14 +50,27 @@ const endpoint: NextApiHandler = async (
...
-          parseInt(req.query.id as string, 10),
-          {
-            ...req.body,
-            createdBy: user.email,
-          }
-        );
-        res.status(StatusCodes.CREATED).json(data);
+        const request_type = req.query.type;
+
+        if (request_type && request_type == 'add_worker_to_allocation') {
```
Comment (acodeninja) [inline pages/api/residents/[id]/allocations/index.ts:55]: Not needed, can be replaced with

```javascript
if (request_type === 'add_worker_to_allocation')
```

File: pages/residents/[id]/allocations/[allocationId]/allocateworker.tsx [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import { isBrowser } from 'utils/ssr';
+import { canUserAllocateWorkerToPerson } from 'lib/permissions';
+import { User } from 'types';
+import { useResident } from 'utils/api/residents';
+import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
+import Spinner from 'components/Spinner/Spinner';
+import React from 'react';
+import AddWorkerAllocation from 'components/Allocations/AddWorkerAllocation/AddWorkerAllocation';
+
+const AddNewAllocationPage = (): React.ReactElement => {
```
Comment (acodeninja) [inline pages/residents/[id]/allocations/[allocationId]/allocateworker.tsx:14]: Maybe rename to `AddWorkerAllocationPage`? 

In this branch there are three pages with the name "AddNewAllocationPage" gets confusing when trying to debug.

**Review by acodeninja — CHANGES_REQUESTED** [review]
A couple of sense check things to resolve but otherwise great work!

File: cypress/integration/team_allocations.ts (hunk 3 above)
Comment (mapleleafjack) [reply cypress/integration/team_allocations.ts:66]: That makes sense,
I'll remove all reference to real names!

File: pages/residents/[id]/allocations/[allocationId]/allocateworker.tsx (hunk 6 above)
Comment (mapleleafjack) [reply pages/residents/[id]/allocations/[allocationId]/allocateworker.tsx:14]: I'll change the name to "AllocateWorkerPage"

**Review by acodeninja — APPROVED** [review]
LGTM

**Pull request merged**